### PR TITLE
Add support for CRAM files

### LINKED ISSUE #9: Allow CRAM as input format
the genomes from 1000genomes consortium are partially only available as CRAMS. It would be useful to allow conversion of them as well. This would require a bit of rewriting of the output channels (i.e. change *.bam) as well as determining how to run FastQC with this input. (see issue: s-andrews/FastQC#54)

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
This aims to resolve #9. I opted for the simple approach of pre-pending the pipeline with a CRAM-to-BAM conversion, which has the advantage of leaving the rest of the pipeline intact. Some folks might want to be more efficient and convert directly to FASTQ, but I figured that it would be useful to have a less efficient implementation in the meantime. 

I wanted to confirm whether this is a welcome change before going through the remaining tasks on the checklist. Thanks! 

Edit: I forgot to mention that this is my first time writing Nextflow (other than trivial config changes), so I would gladly accept any feedback you might have! 

## PR checklist
 - [x] This comment contains a description of changes (with reason)
 - [x] If you've fixed a bug or added code that should be tested, add tests!
 - [x] If necessary, add test data to `testdata/` 
 - [x] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [x] Make sure your code lints (`nf-core lint .`). 
 - [x] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated

**Learn more about contributing:** https://github.com/qbic-pipelines/bamtofastq/tree/master/.github/CONTRIBUTING.md

### CODE REVIEW

File: conf/test_cram.config [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  config_profile_description = 'Minimal test dataset to check pipeline function'
+  // Limit resources so that this can run on Travis
+  max_cpus = 2
+  max_memory = 6.GB
+  max_time = 48.h
+
+  cram_files = true
+  input = [
+          'https://raw.githubusercontent.com/BrunoGrandePhD/bamtofastq/master/testdata/First_SmallTest_Paired.cram',
+          'https://raw.githubusercontent.com/BrunoGrandePhD/bamtofastq/master/testdata/Second_SmallTest_Paired.cram'
```
Comment (BrunoGrandePhD) [inline conf/test_cram.config:22]: This could be replaced with URLs from the main repo immediately prior to merging.

File: main.nf [hunk 2]
```diff
@@ -203,6 +220,53 @@ process get_software_versions {
...
+
+  if ( params.reference_fasta ) {
+
+    process cramToBamWithReference {
+      tag "$name"
+      label 'process_medium'
+
+      input:
+      set val(name), file(cram) from ch_cram_files
+      path fasta from params.reference_fasta
```
Comment (BrunoGrandePhD) [inline main.nf:237]: I tried setting the `fasta` input conditionally, but I couldn't get it to work, so I'm conditionally including the process. Let me know if there's a more elegant approach here.

Comment (BrunoGrandePhD) [reply main.nf]: Never mind. I stumbled across an [approach](https://github.com/nf-core/sarek/blob/2.7.1/main.nf#L2037) in Sarek, which I was able to use here to de-duplicate the processes.

File: main.nf [hunk 3]
```diff
@@ -203,6 +222,32 @@ process get_software_versions {
     """
 }
 
+/*
+ * Generate BAM files if input files are CRAM files
+ */
+
+if ( params.cram_files ) {
+
+  process cramToBamWithReference {
```
Comment (FriederikeHanssen) [inline main.nf:231]: I think it should be possible to use samtolls fastq directly with cram files. http://www.htslib.org/doc/samtools-fasta.html Possibly not easy in DSL1, in that case could leave it as is and update once I get to convert this to DSL2

**Comment by FriederikeHanssen** [comment]
❤️ yes more than welcome addition. Thank you for contributing 😊 I agree, probably easiest to just prepend the process. I am trying to find time to convert this to DSL2, it might be easier to then use `samtolls fastq` directly.

**Comment by FriederikeHanssen** [comment]
nf-core lint might be an issue tbh given the recent release for dsl2, this pipeline doesn't adhere to it anymore. So if linting fails because of DSL2 things, don't worry about it

File: conf/test_cram.config (hunk 1 above)
Comment (FriederikeHanssen) [reply conf/test_cram.config:22]: sounds good, this is how we usually have done it. Bit of a chicken and egg problem.

**Comment by BrunoGrandePhD** [comment]
I've updated the target branch to `dev` and resolved the small merge conflict that arose from that. 

I also installed nf-core tools v1.14 to lint using the pre-DSL2 version. I had to disable all checks related to the pipeline schema since it doesn't exist for `bamtofastq`. You can compare the summaries from linting my branch (`bgrande/cram-support`) with the results from linting the `dev` branch. As you can see, my changes didn't introduce any additional warnings or failures. In fact, two more tests pass. ¯\_(ツ)_/¯

```
❯ nf-core --version
nf-core, version 1.14

❯ cat .nf-core-lint.yml 
schema_lint: False
schema_params: False
schema_description: False

# Run on bgrande/cram-support
❯ nf-core lint .
 [...]
╭───────────────────────╮
│ LINT RESULTS SUMMARY  │
├───────────────────────┤
│ [✔] 111 Tests Passed  │
│ [?]  14 Tests Ignored │
│ [!]  12 Test Warnings │
│ [✗]  36 Tests Failed  │
╰───────────────────────╯

# Run on upstream/dev
❯ nf-core lint .
 [...]
╭───────────────────────╮
│ LINT RESULTS SUMMARY  │
├───────────────────────┤
│ [✔] 109 Tests Passed  │
│ [?]  14 Tests Ignored │
│ [!]  12 Test Warnings │
│ [✗]  36 Tests Failed  │
╰───────────────────────╯
```

**Comment by FriederikeHanssen** [comment]
Looks great. I have triggered the tests now. Onnce they pass I think we are good to go with merging

**Comment by BrunoGrandePhD** [comment]
Okay, I went through the rest of the checklist and documented the new options for CRAM file support. Let me know if you have any suggested changes! I tentatively included this change as part of the `1.2.0` release. Feel free to change this as you see fit. 

Thanks for sharing this workflow!

**Comment by FriederikeHanssen** [comment]
looks great. Like your selection of release name :D For the test file, I would propose we merge this PR now, and then update the path once its merged in an additional PR

**Comment by BrunoGrandePhD** [comment]
I've opened two follow-up PRs to update the paths of the test CRAM files:

- #37 
- #38

**Pull request merged**